### PR TITLE
[SelectOpt] Add handling for not conditions.

### PR DESCRIPTION
**Pull request by davemgreen**
This patch attempts to help the SelectOpt pass detect select groups made up of conditions and not(conditions). Usually these are canonicalized in instcombine to remove the not and invert the true/false values, but this will not happen for Loginal operations, which can be beneficial to convert if they are part of a larger select group. The handling for not's are mostly handled in the SelectLike, which can be marked as Inverted in order to reverse the TrueValue and FalseValue.

This helps fix a regression in fortran minloc constructs, after #84628 helped simplify a loop with branches into a loop with selects.

### CODE REVIEW

**Review by sjoerdmeijer — COMMENTED** [review]
> Usually these are canonicalized in instcombine to remove the not and invert the true/false values, but this will not happen for Loginal operations

Just wanted to quickly check the reason instcombine isn't doing canonicalisations for these cases. I guess it's because the limited scope that instcombine has, so it won't trigger, is that right?

**Comment by davemgreen** [comment]
Thanks

> Just wanted to quickly check the reason instcombine isn't doing canonicalisations for these cases. I guess it's because the limited scope that instcombine has, so it won't trigger, is that right?

It will be because `select x, true, y` is treated as `or x, y` (but poison blocking), and it does not want to break that canonical form.

**Review by sjoerdmeijer — APPROVED** [review]
Cheers, LGTM

File: llvm/lib/CodeGen/SelectOptimize.cpp [hunk 1]
```diff
@@ -153,14 +157,22 @@ class SelectOptimizeImpl {
     bool isValid() { return I; }
     operator bool() { return isValid(); }
 
+    /// Invert the select by inverting the condition and switching the operands.
+    void setInverted() {
+      assert(!Inverted && "Trying to invert and inverted SelectLike");
```
Comment (sapostolakis) [inline llvm/lib/CodeGen/SelectOptimize.cpp:162]: invert and inverted --> invert an inverted

File: llvm/lib/CodeGen/SelectOptimize.cpp [hunk 2]
```diff
@@ -177,11 +189,22 @@ class SelectOptimizeImpl {
       llvm_unreachable("Unhandled case in getCondition");
     }
 
+    /// Return the condition for the SelectLike instruction. For example the
+    /// condition of a select or c in `or(zext(c), x)`
+    Value *getCondition() const {
+      Value *CC = getNonInvertedCondition();
+      if (Inverted)
+        return cast<Instruction>(CC)->getOperand(0);
```
Comment (sapostolakis) [inline llvm/lib/CodeGen/SelectOptimize.cpp:197]: add a comment on what type of instruction can be inverted and why it is safe/correct to get the first operand for the condition

**Review by sapostolakis — APPROVED** [review]
LGTM. I only had two minor comments.

**Comment by davemgreen** [comment]
Thanks

**Pull request merged**